cpp/src/community/egonet_impl.cuh

```diff
@@ -92,67 +90,74 @@ extract(raft::handle_t const& handle,
   hr_timer.start("ego_neighbors");
 #endif
 
-  for (vertex_t i = 0; i < n_subgraphs; i++) {
+  // FIXME: Explore the performance here.  Single-seed BFS
+  // has a slow parallelism ramp up, would we be better off
+  // using the technique in induced subgraph where we tag
+  // the vertices and search for matches until the frontiers
+  // are large enough to use this approach?
+
+  for (size_t i = 0; i < source_vertex.size(); i++) {
     // get light handle from worker pool
     raft::handle_t light_handle(handle.get_next_usable_stream(i));
-    auto worker_stream_view = light_handle.get_stream();
+    auto worker_stream_view = multi_gpu ? handle.get_stream() : light_handle.get_stream();
 
     // BFS with cutoff
     // consider adding a device API to BFS (ie. accept source on the device)
-    rmm::device_uvector<vertex_t> predecessors(v, worker_stream_view);  // not used
     bool direction_optimizing = false;
     thrust::fill(rmm::exec_policy(worker_stream_view),
                  reached[i].begin(),
                  reached[i].end(),
                  std::numeric_limits<vertex_t>::max());
-    thrust::fill(
-      rmm::exec_policy(worker_stream_view), reached[i].begin(), reached[i].begin() + 100, 1.0);
-
-    cugraph::bfs<vertex_t, edge_t, weight_t, false>(light_handle,
-                                                    csr_view,
-                                                    reached[i].data(),
-                                                    predecessors.data(),
-                                                    source_vertex + i,
-                                                    1,
-                                                    direction_optimizing,
-                                                    radius);
+
+    cugraph::bfs<vertex_t, edge_t, weight_t, multi_gpu>(multi_gpu ? handle : light_handle,
+                                                        graph_view,
+                                                        reached[i].data(),
+                                                        nullptr,
+                                                        source_vertex.data() + i,
+                                                        1,
+                                                        direction_optimizing,
+                                                        radius,
+                                                        do_expensive_check);
```

Comment:
Maybe just add a FIXME and address later, but this looks very inefficient (but not less efficient than what we had... so better kick out cython.cu first).

Sequentially running one BFS (with maximum distance of radius) per source vertex can largely under-utilize a GPU (or multiple GPUs in multi-GPU runs).

We may better either run multi-source BFS, or run 1-hop neighbors recursively (radius times).

Reply:
There is a FIXME on line 93 (about 20 lines higher) observing that this is likely slow.  Is that sufficient to address this concern?

Reply:
I saw that FIXME but wasn't sure about the `until the frontiers are large enough to use this approach` part. But we can defer this discussion to the future.

The reason I wasn't sure about the FIXME comment.

I think we'd better not use `this approach` (i.e. running single source BFS in a sequential loop). Running 1-hop neighbors recursively sounds like a simpler approach with optimal parallelism but we may revisit this in the future when we actually update this algorithm.